compiler/src/iree/compiler/Codegen/LLVMCPU/Passes.cpp

```diff
@@ -52,6 +52,12 @@ static llvm::cl::opt<bool> clEnableMicrokernels(
     llvm::cl::desc("Enables microkernel lowering for vmvx (experimental)"),
     llvm::cl::init(false));
 
+static llvm::cl::opt<bool> clEnableReductionReordering(
+    "iree-llvmcpu-enable-reduction-reordering",
+    llvm::cl::desc("Enables reordering reductions for floats that are not"
+                   "(strictly) associative for addition or multiplication"),
+    llvm::cl::init(true));
```

Comment:
I thought the idea was to enable reassociation by default, didn't it?

Reply:
Hmm, the default should be to take this path right? Why should the default be false?

Reply:
ah yes, I missed that we've checked the element types should be integers. In the integer case, it should be default on. Sorry about that.

For the floating point, it should be default off, and have a flag to control it right?

Reply:
Can we resolve the conversation only when there are no other questions or the issue is addressed... I still don't get the answer about FP yet...

I do have concerns if it's default on for FP. I did spend couple weeks on debugging numerical issues because of a similar feature. IMO, we should not enable the features for FP by default and should have a flag for controlling it. People who care about performance (more than numerical) can enable the feature themselves. LLVM also has the same restriction. It's off by default, and it's on in fast-math mode. https://clang.llvm.org/docs/UsersManual.html#id12

I assumed that the flag is for controlling the behavior for FP, but it seems not to me now.. It's very confusing to me because the description is about enabling the features for FP or not.

Reply:
Okay I will let you resolve the conversation from now on. I just use resolve whenever I think it's resolved or I have made code changes to fix the issue usually.

> I do have concerns if it's default on for FP. I did spend couple weeks on debugging numerical issues because of a similar feature. IMO, we should not enable the features for FP by default and should have a flag for controlling it. People who care about performance (more than numerical) can enable the feature themselves. LLVM also has the same restriction. It's off by default, and it's on in fast-math mode. https://clang.llvm.org/docs/UsersManual.html#id12

Okay. I will make it off by default.

> I assumed that the flag is for controlling the behavior for FP, but it seems not to me now.. It's very confusing to me because the description is about enabling the features for FP or not.

Yes it is. The feature is turned on by default for all integer types. For any other type (FP included) it is controlled by the flag

Reply:
Yes, let's rename this flag to `iree-llvmcpu-enable-fp-reduction-reassoc` or similar.

Regarding its default value, I'm ok with setting it up to `false` by default if there are concerns. However, we have to be consistent. Are we using precise implementations for all the arithmetic and math operations subject to leading to numerical differences (e.g., div, rem, sqrt, exp, log, etc.? What are the reference TFLite implementations doing? Perhaps this is something to bring up with the rest of the team.

Reply:
> Are we using precise implementations for all the arithmetic and math operations subject to leading to numerical differences (e.g., div, rem, sqrt, exp, log, etc.?

Yes, we are.

> What are the reference TFLite implementations doing? Perhaps this is something to bring up with the rest of the team.

I don't know what TFLite implements is. I know that there are many flags in reference tensorflow implementation. When doing correctness tests, we disable aggressive flags. (one of the examples: https://github.com/iree-org/iree/issues/9853#issuecomment-1191913120)

My main point is that given a dispatch or a model, all the backends in IREE should generate identical results by default (for FP cases, the accuracy could be <= 1e-6). Turning the feature on by default will make IREE backends get diverged. The fancy optimization should be controlled under a flag (say iree-fast-math) if they could impact accuracy a lot.

Reply:
Added a flag `--iree-llvmcpu-enable-reduction-reordering-for-all`

Reply:
> My main point is that given a dispatch or a model, all the backends in IREE should generate identical results by default (for FP cases, the accuracy could be <= 1e-6). Turning the feature on by default will make IREE backends get diverged. The fancy optimization should be controlled under a flag (say iree-fast-math) if they could impact accuracy a lot.

I agree with the idea. I just want to point out that this isn't how it is on the GPU backends (both llvmgpu and spirv) right now as not distributing reduction makes performance collapse. This is also true for most framework out there. I agree that this is unfortunate and it would be nice to have a mode with strict floating points but I don't think we will be able to afford to not do this kind of re-association for reductions by default.

Reply:
> I just want to point out that this isn't how it is on the GPU backends (both llvmgpu and spirv) right now as not distributing reduction makes performance collapse. This is also true for most framework out there. 

Are you implying that it would diverge more if this is not turned on?

Reply:
not really, I just meant this is necessary for some targets. Every target is likely to do partial accumulation with a different tile size so we can't really bet on all the target doing re-association and expect the same result.

Reply:
> I doubt we can diverge that much by reordering reductions. But nevertheless I have made the flag false by default.

Here is an example that the difference increased from 1e-6 to 1e-2 because of re-association optimization. https://github.com/iree-org/iree/issues/9536

We could see more examples when connecting more models e2e.

I'm surprised that CPU and GPU backends are diverged today. :o 
IMO, we should control these fancy optimizations under one flag. I can't speak if they will be on by default in the future, but having them under one flag controlled and having ability to generate "golden" values are important. It's also easier if we really want to flip to that pass by default in the future.

Reply:
> Here is an example that the difference increased from 1e-6 to 1e-2 because of re-association optimization. https://github.com/iree-org/iree/issues/9536

Interesting. Looks like there's a F64 to F32 conversion as well. Is that orthogonal to the rounding bug?

Reply:
> I'm surprised that CPU and GPU backends are diverged today. :o
> IMO, we should control these fancy optimizations under one flag. I can't speak if they will be on by default in the future, but having them under one flag controlled and having ability to generate "golden" values are important. It's also easier if we really want to flip to that pass by default in the future.

I completely agree there should be a flag but it would have to be a per operation flag to be meaningful. I believe Jakub has started some discussion on discourse about fast math but this is going to be a non trivial change.

> Interesting. Looks like there's a F64 to F32 conversion as well. Is that orthogonal to the rounding bug?

I think this is quite different.converting F64 to F32 is usually plain wrong.

Reply:
Thomas, thanks for the clarification, it really help!

> Interesting. Looks like there's a F64 to F32 conversion as well. Is that orthogonal to the rounding bug?

That is a different issue. The root cause is re-association. https://github.com/iree-org/iree/issues/9536#issuecomment-1164844079

Anyway, it's just an example issue that I've hit about re-association in IREE. I'm happy to see the pass on for int by default, and off for FP now. Thanks for pushing this forward, it's an awesome work!

Reply:
Yeah, reduction error can surprisingly accumulate and lead to major output differences. I agree with the idea of having a higher precision mode but I'm not sure if the bit-exact goal across targets is feasible. Different targets have different instructions for fundamental operations using different approximations which would lead to mismatching precision. We would have to use the same high-accuracy software-based emulation to get something along those lines. Not sure how realistic this is.

Reply:
Can we resolve this comment?

Reply:
The issue is addressed, but I'd like to keep the conversation open. I found that keeping this kind of conversation open is helpful when people play something similar and could lead them to the PR to see some discussions.

Reply:
Just my opinion here... I don't like to resolve the comments because it hides the discussion and it's difficult for reviewers to check if the comments were addressed or not. As a reviewer, you have to go and open all the comments one by one. To mark comments as addressed for myself, I use an emoji and that usually works well for me. Of course, you can do whatever works better for you :)